src/library/scala/collection/immutable/LazyList.scala

```diff
@@ -504,18 +490,35 @@ final class LazyList[+A] private(private[this] var lazyState: () => LazyList.Sta
     */
   override def collect[B](pf: PartialFunction[A, B]): LazyList[B] =
     if (knownIsEmpty) LazyList.empty
-    else collectTrampoline(pf.lift)
+    else LazyList.collectImpl(this, pf)
 
-  // trampoline to allow for tail-recursive `collectState`
-  @inline private def collectTrampoline[B](lifted: A => Option[B]): LazyList[B] =
-    newLL(collectState(lifted))
+  /** @inheritdoc
+    *
+    * This method does not evaluate any elements further than
+    * the first element for which the partial function is defined.
+    */
+  @tailrec
+  override def collectFirst[B](pf: PartialFunction[A, B]): Option[B] =
+    if (isEmpty) None
+    else {
+      val marker = Statics.pfMarker
+      val res = pf.applyOrElse(head, ((_: A) => marker).asInstanceOf[A => B])
+      if (res.asInstanceOf[AnyRef] eq marker) tail.collectFirst(pf)
```

Comment:
While this does get translated to a tail call, you end up fetching `pfMarker` and allocating the `(_: A) => marker)` closure for every element.

Reply:
That's a good point. I can definitely cache the function in the companion.

What is the cost of fetching `pfMarker` repeatedly? And if I cache the function as `(_: Any) => Statics.pfMarker`, does that also fetch it each time it invokes the function? Is `val marker = Statics.pfMarker; (_: Any) => marker` better or worse?

Reply:
@szeiger I can't figure out how to make the head collectible without a tail call. If you turn the implementation in the class into a while loop, then it's leaked because it's still the head doing the execution. And if you have it forward to a static method in the companion, then it appears to be leaked because the parent stack frame references the head (because it called `collectFirst` on it).

Reply:
it seems like the generated bytecode for a tail call overwrites `this` (see [this comment](https://github.com/scala/bug/issues/11443#issuecomment-481493790) from Seth), and I'm not sure how (or if it's possible) to replicate that without a tail call.

Reply:
Right, there's no good option at the moment. We previously did some hacks with extension methods but that doesn't help when you have to implement an interface. I've been thinking about an annotation-based optimization for these cases (see my comments on Gitter) but that's not an option for 2.13.0. Something close to your current implementation is probably the best we can get.

I'm not sure if the closure allocation can be hoisted out of the loop by the optimizer if it doesn't reference `marker`. It's worth a try but I doubt it. I'd reference the marker directly from inside the closure instead of copying it into a local val first. I've done that in other places because it's cheaper when you have to read it multiple times but that doesn't apply if we can't get it out of the loop.

Reply:
OK, looking at the other implementations again: What about `filter`? It delegates to a static method. Are you saying that this will drop the receiver reference but the same doesn't work for `collectFirst`?

Reply:
Oh, I see. `filter` works because it's not a terminal method. `filter` does keep a reference to the receiver but it's not a problem but it simply returns a new LazyList. The actual filtering is done later.